Re-work test cluster for Babbage integration tests

Anviking opened the pull request:
### Overview

(see commit messages)

- Refactor cluster setup
- Add support for Babbage, where we register pools in genesis, rather than via txs to a BFT node. BFT nodes don't work in Babbage.


### Comments

- Nightly https://buildkite.com/input-output-hk/cardano-wallet-nightly/builds/1416#0181f36a-b8ef-4b1f-9fa1-f4fe5becae77 (latency bench still works)

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

https://input-output.atlassian.net/browse/ADP-1815

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

Anviking committed:
Fix icarus address discovery in genesis block

This matters when we fund wallets via genesis in our new integration test
setup.

Anviking committed:
Remove NO_POOLS env var in cluster

It was useful for making the cluster startup shorter by not waiting
multiple epochs for the pools to get stake. With babbage we can register
pools in genesis to be active immediately, so the primary use for NO_POOLS
is removed.

In alonzo or prior eras, NO_POOLS could be useful, but for simplicity
let's remove it anyway.

Anviking committed:
Make pool follower able to detect genesis pools

This will allow integration tests to use the ShelleyGenesis `sgStaking`
field for the pools.

Anviking committed:
Refactor cluster setup

Prepare for the addition of babbage support.

- Prefer adding initial funds via shelley genesis rather than byron
  genesis or tx submission.
  - This could break large wallets.
- Tweak epoch length from 100 to 80 (not sure why but going with it)
- Make cluster setup clearer and a bit more strongly typed with
  `GenesisFiles`, `PoolRecipe`, `ConfiguredPool`, `configurePool` etc.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -727,6 +729,12 @@ monitorStakePools tr (NetworkParameters gp sp _pp) nl DBLayer{..} =
             toChainPoint (BlockHeader sl _ h _) = ChainPoint sl h
 
+
+        let psudoGenesisSlotNo = SlotNo 0
+        atomically $ do
+            putPoolCertificates psudoGenesisSlotNo genesisPools
```
This may be sketchy (what happens if you restart the wallet?), but is only used in integration tests, and deserves a notice ⚠️

Anviking commented on lib/shelley/test/data/cardano-node-shelley/node.config:
```diff
@@ -14,8 +14,14 @@ RequiresNetworkMagic: RequiresNoMagic
+# The maximum number of used peers during bulk sync.
+MaxConcurrencyBulkSync: 1
+# The maximum number of used peers when fetching newly forged blocks.
+MaxConcurrencyDeadline: 2
```
Not sure if we need all this... Probably we need `MaxConcurrency...` but not sure about the rest...

Anviking replied:
Dropped the rest

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -1059,8 +1060,8 @@ spec = describe "SHELLEY_STAKE_POOLS" $ do
                     `shouldBe`
                     Set.fromList
-                        [ Quantity oneMillionAda
-                        , Quantity $ 2 * oneMillionAda
+                        [ Quantity $ 100 * oneMillionAda
+                        , Quantity $ 200 * oneMillionAda
```
This is apart of the wrong commit.

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/TransactionsNew.hs:
```diff
@@ -1765,17 +1766,18 @@ spec = describe "NEW_SHELLEY_TRANSACTIONS" $ do
             (Link.balanceTransaction @'Shelley wa) Default toBalance'
+
+        let (requiredAmt, largestFound) =
+                -- May need babbage specific values when
```
Meant to say when "we address ADP-1909"

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -1430,10 +1431,18 @@ spec = describe "SHELLEY_STAKE_POOLS" $ do
+        if _mainEra ctx >= ApiBabbage
+        then costOf (\coeff cst -> 458 * coeff + cst) ctx
+         -- FIXME: with 2 more bytes in babbage this test still failed:
+         --  STAKE_POOLS_JOIN_01x - Fee boundary values, STAKE_POOLS_JOIN_01x
```
Right, why is this 4 bytes? 2 bytes makes sense (if outputs were 2 bytes bigger on babbage), but I don't know why there are 2 more. Most likely nothing to worry about, but should replace the `FIXME` with a `NOTE`.

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Byron/Transactions.hs:
```diff
@@ -493,7 +493,7 @@ spec = describe "BYRON_TRANSACTIONS" $ do
         verify r
             [ expectResponseCode HTTP.status200
-            , expectListSize 1 -- Now funded through a tx in the cluster setup
+            , expectListSize 10
```
In the wrong commit - related to refactoring, not to the addition of babbage support

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:
```diff
@@ -231,7 +232,7 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
             r <- request @([ApiTransaction n]) ctx link Default Empty
             expectResponseCode HTTP.status200 r
-            expectListSize 1 r
+            expectListSize 10 r
```
Wrong commit

Anviking commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:
```diff
@@ -2195,6 +2196,11 @@ spec = describe "SHELLEY_TRANSACTIONS" $ do
 
+        when (_mainEra ctx < ApiBabbage) $
+            liftIO $ pendingWith
+                "expected fees have been updated to Babbage and these \
+                \tests marked pending on earlier eras."
```
Quick solution to avoid having to make `CreateTransactionWithMetadataTest` more complicated

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs:
```diff
@@ -892,31 +949,44 @@ withCluster tr dir LocalClusterConfig{..} initialFunds onClusterStart = bracketT
-            -- NOTE: We used to perform 'registerViaTx' as part of 'launchPools'
-            -- where we waited for the pools to become active (e.g. get a stake
-            -- distribution)  in parallel. Now waiting seems to work fine
-            -- though, and we will soon only run tests on babbage anyway...
```
NOTE shouldn't be removed in the babbage commit

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs:
```diff
@@ -892,31 +949,44 @@ withCluster tr dir LocalClusterConfig{..} initialFunds onClusterStart = bracketT
-                cfgNodeLogging
-        withBFTNode tr dir bftCfg $ \bftSocket block0 param -> do
-            -- FIXME: genesis pools may techincally be incorrect;
-            -- withBFTNode should pass in the @RunningNode@ instead.
```
Might as well fix

Anviking committed:
Add Babbage integration test support

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs:
```diff
@@ -537,77 +878,113 @@ withCluster
-                        (port0, ports) cfgNodeLogging
-                withRelayNode tr dir cfg $ \socket -> do
-                    let runningRelay = RunningNode socket block0 params
-                    onClusterStart runningRelay `finally` cancelAll
```
I think I missed this `finally cancelAll`, causing CI timeouts 😱

Anviking commented on lib/shelley/test/integration/shelley-integration-test.hs:
```diff
@@ -345,6 +359,7 @@ specWithServer testDir (tr, tracers) = aroundAll withContext
                 gp
                 tunedForMainnetPipeliningStrategy
                 (SomeNetworkDiscriminant $ Proxy @'Mainnet)
+                genesisPools
```
This diff is in the wrong commit... but might not bother fixing in this case.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs:
```diff
@@ -427,27 +421,253 @@ cliRetry tr msg processConfig = do
       -- ^ An optional retirement epoch. If specified, then a pool retirement
       -- certificate will be published after the pool is initially registered.
+    , poolMetadata :: Aeson.Value
+    , delisted :: Bool
```
Currently has no effect

Anviking replied:
Now fixed 1e16fbfa1b1038a68e4242477d54e8671543bd16

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs:
```diff
@@ -427,27 +421,253 @@ cliRetry tr msg processConfig = do
+                    --
+                    -- NOTE: The retirement is duplicated with the one in
+                    -- @registerViaTx@. That might still work, and we might want
+                    -- to remove @registerViaTx@.
```
Should rephrase

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs:
```diff
@@ -516,6 +740,123 @@ data RunningNode = RunningNode
     Block
     -- ^ Genesis block
     (NetworkParameters, NodeToClientVersionData)
+    [PoolCertificate]
+    -- ^ Shelley genesis pools
```
Wrong commit, probably

Anviking committed:
Unify operatorsFixture and defaultPoolConfigs

sevanspowell commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs:
```diff
@@ -427,27 +391,252 @@ cliRetry tr msg processConfig = do
+        , delisted = True
+        , index = undefined
+        }
+    -- This pool should retire almost immediately:
```
```suggestion
    -- This pool should retire, but not within the duration of a test run:
```

sevanspowell commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs:
```diff
@@ -427,27 +391,252 @@ cliRetry tr msg processConfig = do
+        , delisted = False
+        , index = undefined
+        }
+
```
```suggestion
    -- This pool should retire almost immediately:
```

sevanspowell commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs:
```diff
@@ -427,27 +391,252 @@ cliRetry tr msg processConfig = do
+configurePools tr dir metadataServer =
+    mapM (configurePool tr dir metadataServer)
+
+configurePool
```
I like the work you've done here with the clearer types and easier-to-understand program flow :100:

sevanspowell commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch/Cluster.hs:
```diff
@@ -427,27 +391,252 @@ cliRetry tr msg processConfig = do
+              , "description" .= Aeson.String "Lorem Ipsum Dolor Sit Amet."
+              , "homepage" .= Aeson.String "https://iohk.io"
+              ]
+        , delisted = True
```
A note here would be good I think. What does `delisted` mean for this pool? Will we be able to interact with it during the test run?

Anviking replied:
e2027d4c51bc90a481344b78a76a1f763f5071b1 - btw there is some integration test that checks that the wallet also lists delisted pools (but with a delisted flag, I think)

sevanspowell commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/NetworkSpec.hs:
```diff
@@ -246,10 +253,13 @@ withTestNode
-testnet :: NetworkId
-testnet = Testnet $ NetworkMagic 0
+mainnet :: NetworkId
+mainnet = Testnet $ NetworkMagic 764824073
```
Hmm, can you explain why this is called "mainnet"?

Anviking replied:
Just using `Mainnet` seems to work as well, so going with that 👍 8dbe98e0469f2b5c1824e8a2ea92bcbd3018ebdf

sevanspowell reviewed (approved):
I have a few suggestions in my comments, but otherwise this looks good to me!

Anviking committed:
Correct comments in defaultPoolConfigs

Co-authored-by: Samuel Evans-Powell <mail@sevanspowell.net>

Anviking committed:
Document `delisted` field

Anviking committed:
Tweak indentation

Anviking committed:
fixup: use `Mainnet` rather than `Testnet magic` in NetworkSpec

Anviking committed:
Add cautionary note about `bigDustWallet`

Anviking committed:
Make maxLengthAddress longer

From integration test failures, it seems the maxLengthAddress was too
short -- Byron addresses can be longer.

This is one such longer address:
```
> let Right (Address addr) = decodeAddress @'Mainnet "DdzFFzCqrht74rkP7eNhMp9iaQ79JQZzHX6QxjoFoie4qAn5D2MESx3Rzpqtc9zX6ASEdDThwJyqjc2kjqHMFnoUnC79GmmNCB9Vfe6a"
> BS.length addr
76
```

However the required additional overestimation seem to cause - partially
unrelated - integration tests to fail.

Anviking committed:
Temporarily use Alonzo calculateMinUTxOValue in Babbage

To allow us to merge the babbage integration tests before resolving the
newly discovered problems with our babbage-specific minUTxOValue
calculation.
